x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java

```diff
@@ -477,37 +475,53 @@ synchronized void handleFailure(Exception e) {
             handleIrrecoverableBulkIndexingException((BulkIndexingException) unwrappedException);
-        } else if (unwrappedException instanceof IndexNotFoundException
-            || unwrappedException instanceof AggregationResultUtils.AggregationExtractionException
-            || unwrappedException instanceof TransformConfigReloadingException
-            || unwrappedException instanceof ResourceNotFoundException
-            || unwrappedException instanceof IllegalArgumentException) {
-                failIndexer("task encountered irrecoverable failure: " + e.getMessage());
-            } else if (context.getAndIncrementFailureCount() > context.getNumFailureRetries()) {
-                failIndexer(
-                    "task encountered more than "
-                        + context.getNumFailureRetries()
-                        + " failures; latest failure: "
-                        + ExceptionRootCauseFinder.getDetailedMessage(unwrappedException)
-                );
-            } else {
-                // Since our schedule fires again very quickly after failures it is possible to run into the same failure numerous
-                // times in a row, very quickly. We do not want to spam the audit log with repeated failures, so only record the first one
-                if (e.getMessage().equals(lastAuditedExceptionMessage) == false) {
-                    String message = ExceptionRootCauseFinder.getDetailedMessage(unwrappedException);
+            return;
+        }
 
-                    auditor.warning(
-                        getJobId(),
-                        "Transform encountered an exception: " + message + " Will attempt again at next scheduled trigger."
-                    );
-                    lastAuditedExceptionMessage = message;
-                }
+        // irrecoverable error without special handling
+        if (unwrappedException instanceof ElasticsearchException) {
+            ElasticsearchException elasticsearchException = (ElasticsearchException) unwrappedException;
+            if (ExceptionRootCauseFinder.IRRECOVERABLE_REST_STATUSES.contains(elasticsearchException.status())) {
+                failIndexer("task encountered irrecoverable failure: " + elasticsearchException.getDetailedMessage());
+                return;
             }
+        }
+
+        if (unwrappedException instanceof IllegalArgumentException) {
+            failIndexer("task encountered irrecoverable failure: " + e.getMessage());
+            return;
+        }
+
+        if (context.getAndIncrementFailureCount() > context.getNumFailureRetries()) {
+            failIndexer(
+                "task encountered more than "
+                    + context.getNumFailureRetries()
+                    + " failures; latest failure: "
+                    + ExceptionRootCauseFinder.getDetailedMessage(unwrappedException)
+            );
+        } else {
```

Comment:
You could `return;` in line 511 and get rid of `else` clauses, just like you did above.